webhook/resourcesemantics/validation/validation_context.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"context"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Context wraps information relevant for validation of the incoming request along with the generic context.Context.
+type Context struct {
+	// CallCtx is the context from the user-request.
+	CallCtx context.Context
+
+	// Operation is the operation being performed. This may be different than the operation
+	// requested. e.g. a patch can result in either a CREATE or UPDATE Operation.
+	Operation admissionv1beta1.Operation
+
+	// DryRun indicates that this request will not be persisted.
+	// Any operations with persistent side-effects should be skipped.
+	DryRun bool
+
+	// KubeClient is a client for interacting with k8s APIs.
+	KubeClient kubernetes.Interface
+}
+
+// NewValidationContext copies relevant objects from the request to a single structure to be read during validation.
+func NewValidationContext(ctx context.Context, ac *reconciler, req *admissionv1beta1.AdmissionRequest) (vc *Context) {
```

Comment:
golang's context is essentially a bag of key => values and deadline/cancellation support.

You don't need to wrap the request context in a new struct but instead decorate the existing one - see: https://golang.org/pkg/context/#example_WithValue & https://github.com/knative/pkg/blob/master/logging/logger.go

The controller also provides a `withContext` so we may only need to decorate this instance with only the request specific things - ie. operation, dry run.

you may not need to need to pass the kubeclient on this context since it should be available on the context that's passed to the webhook controllers [here](https://github.com/knative/serving/blob/ee26e64f2158112933834945620477af540efc77/cmd/webhook/main.go#L106).

Reply:
From: https://github.com/knative/pkg/issues/1140#issuecomment-598877765

>It might be useful to decorate the context with whether we're in a DryRun or not. e.g. webhook.IsInDryRun(ctx), so that things can know and act accordingly.

Reply:
I've removed the wrapper in favor of more parameters for clarity.